File: Assets/KiwiFramework/Core/PMVC/NotifySyetem/NotifySystem.cs
using System;
using KiwiFramework.Core.Interface;


namespace KiwiFramework.Core
{
    /// <summary>
    /// 全局通知系统
    /// </summary>
    public static class NotifySystem
    {
        /// <summary>
        /// 通知中心
        /// </summary>
        private static readonly NotifyCenter<string, INotice> Center = new NotifyCenter<string, INotice>();

        /// <summary>
        /// 发送通知
        /// </summary>
        /// <param name="msgCode">消息号</param>
        /// <param name="notice">消息数据</param>
        public static void Send(string msgCode, INotice notice)
        {
            Center.Send(msgCode, notice);
        }

        /// <summary>
        /// 添加通知监听
        /// </summary>
        /// <param name="msgCode">消息号</param>
        /// <param name="function">通知执行事件</param>
        public static void AddListener(string msgCode, Action<INotice> function)
        {
            Center.Listen(msgCode, function);
        }

        /// <summary>
        /// 移除通知监听
        /// </summary>
        /// <param name="msgCode">消息号</param>
        /// <param name="function">通知执行事件</param>
        public static void RemoveListener(string msgCode, Action<INotice> function)
        {
            Center.RemoveListen(msgCode, function);
        }

        /// <summary>
        /// 移除所有监听
        /// </summary>
        /// <param name="msgCode">消息号</param>
        public static void RemoveAllListeners(string msgCode)
        {
            Center.RemoveCommand(msgCode);
        }

        /// <summary>
        /// 清除所有监听
        /// </summary>
        public static void ClearAll()
        {
            Center.ClearAll();
        }
    }
}

File: Assets/GameAssets/Luas/PMVC/NotifySystem/NotifySystem.lua
---
--- NotifySystem
---

---@class NotifySystem
local _M = { _VERSION = "1.0" }

---@type NotifyCenter
local NotifyCenter = require("NotifyCenter")

local log = Log.Create("NotifySystem")
----------------------------------------------------------------------------------

local CenterName = { Main = "Main" }

---@type table<string,NotifyCenter>
local centerMap = {
    Main = NotifyCenter:new(CenterName.Main)
}

---添加消息通知中心
---@param centerName string 通知中心名称
function _M.AddNotifyCenter(centerName)
    CenterName = CenterName or {}
    CenterName[centerName] = centerName
    centerMap = centerMap or {}
    centerMap[centerName] = NotifyCenter:new(CenterName[centerName])
end

---获取通知中心
---@param centerName string 中心名称
---@return NotifyCenter
local function GetNotifyCenter(centerName)
    if centerName == nil then
        centerName = CenterName.Main
    end
    return centerMap[centerName]
end

---发送通知
---@param notifyCode NotifySystem.Code 消息号
---@param data table 消息所携带的数据
---@param centerName string 消息中心名称
function _M.Send(notifyCode, data, centerName)
    GetNotifyCenter(centerName):Send(notifyCode, data)
end

---添加监听
---@param notifyCode NotifySystem.Code 消息号
---@param method NotifySystem.Function 通知事件
---@param centerName string 消息中心名称
function _M.AddListener(notifyCode, method, centerName)
    GetNotifyCenter(centerName):Listen(notifyCode, method)
end

---移除监听
---@param notifyCode NotifySystem.Code 消息号
---@param method NotifySystem.Function 通知事件
---@param centerName string 消息中心名称
function _M.RemoveListener(notifyCode, method, centerName)
    GetNotifyCenter(centerName):RemoveListen(notifyCode, method)
end

---移除指定 Block 下的所有监听
---@param Block string 消息 Block
---@param centerName string 消息中心名称
function _M.RemoveListenerForBlock(Block, centerName)
    GetNotifyCenter(centerName):RemoveListenForBlock(Block)
end

---移除指定消息 Code 下的所有监听
---@param notifyCode NotifySystem.Code 消息号
---@param centerName string 消息中心名称
function _M.RemoveListenerForCode(notifyCode, centerName)
    GetNotifyCenter(centerName):RemoveListenForCode(notifyCode)
end

---清除指定消息中心内的所有监听
---@param centerName string 消息中心名称
function _M.RemoveAllListener(centerName)
    GetNotifyCenter(centerName):ClearAll()
end

---清除所有通知中心的所有监听
function _M.ClearAll()
    for _, center in pairs(centerMap) do
        center:ClearAll()
    end
end

return _M